starlette/middleware/http.py

```diff
@@ -0,0 +1,81 @@
+from typing import AsyncGenerator, Callable, Optional, Union
+
+from .._compat import aclosing
+from ..datastructures import MutableHeaders
+from ..responses import Response
+from ..types import ASGIApp, Message, Receive, Scope, Send
+
+_HTTPDispatchFlow = Union[
+    AsyncGenerator[None, Response], AsyncGenerator[ASGIApp, Response]
+]
```

Comment:
```suggestion
_HTTPDispatchFlow = Union[
    AsyncGenerator[None, Response], AsyncGenerator[Response, Response]
]
```

Just because we _can_ accept an arbitrary ASGI app doesn't mean we should. I think we should go with the more restrictive version for now because:
1. A user that wants an arbitrary ASGI app can just add a `type: ignore` and it will work
2. It's easier to add an API than remove it. We can loosen the type annotation if/when someone asks for it. This way we can assume it's a `Response` if we want to use that to implement some other feature down the road.

Reply:
Sounds fair!